File: backend/app/services/note_service.py
import json
import shutil
import uuid
from datetime import UTC, datetime
from pathlib import Path

from fastapi import HTTPException

from app.schemas.note import Note, NoteCreate, NoteDetail, NoteListItem
from app.services.workspace_service import get_workspace_dir

METADATA_FILENAME = "notes.jsonl"
SNIPPET_LENGTH = 140


def _metadata_path(workspace_dir: Path) -> Path:
    return workspace_dir / METADATA_FILENAME


def _note_path(workspace_dir: Path, note_id: str) -> Path:
    return workspace_dir / "notes" / f"{note_id}.md"


def _read_all(workspace_dir: Path) -> list[dict]:
    path = _metadata_path(workspace_dir)
    if not path.exists():
        return []
    return [json.loads(line) for line in path.read_text().splitlines() if line.strip()]


def _write_all(workspace_dir: Path, entries: list[dict]) -> None:
    _metadata_path(workspace_dir).write_text("".join(json.dumps(entry) + "\n" for entry in entries))


def _sort(entries: list[dict]) -> list[dict]:
    entries = sorted(entries, key=lambda e: e["updated_at"], reverse=True)
    entries.sort(key=lambda e: e["pinned"], reverse=True)
    return entries


def _find(entries: list[dict], note_id: str) -> dict:
    # note_id is only ever looked up against this per-workspace index, which
    # only ever contains ids we generated ourselves — so an attacker-supplied
    # note_id that doesn't correspond to a real note simply 404s here, before
    # it can be used to build a filesystem path (unlike workspace_id, which
    # has no such index and needs its own slug regex check).
    for entry in entries:
        if entry["id"] == note_id:
            return entry
    raise HTTPException(status_code=404, detail="Note not found")


def _to_note(entry: dict) -> Note:
    return Note(**entry)


def _make_snippet(content: str, query: str) -> str:
    content = " ".join(content.split())
    if not content:
        return ""
    if not query:
        return content[:SNIPPET_LENGTH]

    idx = content.lower().find(query.lower())
    if idx == -1:
        return content[:SNIPPET_LENGTH]

    start = max(0, idx - SNIPPET_LENGTH // 3)
    end = min(len(content), start + SNIPPET_LENGTH)
    prefix = "…" if start > 0 else ""
    suffix = "…" if end < len(content) else ""
    return f"{prefix}{content[start:end]}{suffix}"


def _to_list_item(workspace_dir: Path, entry: dict, query: str = "") -> NoteListItem:
    content = _note_path(workspace_dir, entry["id"]).read_text()
    return NoteListItem(**entry, snippet=_make_snippet(content, query))


def note_exists(workspace_id: str, note_id: str) -> bool:
    workspace_dir = get_workspace_dir(workspace_id)
    return any(e["id"] == note_id for e in _read_all(workspace_dir))


def list_notes(
    workspace_id: str, favorite_only: bool = False, pinned_only: bool = False
) -> list[NoteListItem]:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _sort([e for e in _read_all(workspace_dir) if not e["trashed"]])
    if favorite_only:
        entries = [e for e in entries if e["favorite"]]
    if pinned_only:
        entries = [e for e in entries if e["pinned"]]
    return [_to_list_item(workspace_dir, e) for e in entries]


def search_notes(
    workspace_id: str, query: str, favorite_only: bool = False, pinned_only: bool = False
) -> list[NoteListItem]:
    query = query.strip()
    if not query:
        return list_notes(workspace_id, favorite_only=favorite_only, pinned_only=pinned_only)

    workspace_dir = get_workspace_dir(workspace_id)
    lower_query = query.lower()
    matches = []
    for entry in _read_all(workspace_dir):
        if entry["trashed"]:
            continue
        if favorite_only and not entry["favorite"]:
            continue
        if pinned_only and not entry["pinned"]:
            continue
        if lower_query in entry["title"].lower():
            matches.append(entry)
            continue
        if lower_query in _note_path(workspace_dir, entry["id"]).read_text().lower():
            matches.append(entry)

    return [_to_list_item(workspace_dir, e, query) for e in _sort(matches)]


def list_recent(workspace_id: str, limit: int = 20) -> list[NoteListItem]:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = [e for e in _read_all(workspace_dir) if e["opened_at"] and not e["trashed"]]
    entries.sort(key=lambda e: e["opened_at"], reverse=True)
    return [_to_list_item(workspace_dir, e) for e in entries[:limit]]


def list_trash(workspace_id: str) -> list[NoteListItem]:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = [e for e in _read_all(workspace_dir) if e["trashed"]]
    entries.sort(key=lambda e: e["trashed_at"], reverse=True)
    return [_to_list_item(workspace_dir, e) for e in entries]


def list_folders(workspace_id: str) -> list[str]:
    workspace_dir = get_workspace_dir(workspace_id)
    folders = {e["folder"] for e in _read_all(workspace_dir) if e["folder"]}
    return sorted(folders)


def list_tags(workspace_id: str) -> list[str]:
    workspace_dir = get_workspace_dir(workspace_id)
    tags: set[str] = set()
    for entry in _read_all(workspace_dir):
        tags.update(entry["tags"])
    return sorted(tags)


def get_note(workspace_id: str, note_id: str) -> NoteDetail:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["opened_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    content = _note_path(workspace_dir, note_id).read_text()
    return NoteDetail(**entry, content=content)


def create_note(workspace_id: str, data: NoteCreate) -> NoteDetail:
    title = data.title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="Note title cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    note_id = uuid.uuid4().hex
    now = datetime.now(UTC).isoformat()
    entry = {
        "id": note_id,
        "title": title,
        "created_at": now,
        "updated_at": now,
        "favorite": False,
        "pinned": False,
        "folder": None,
        "tags": [],
        "trashed": False,
        "trashed_at": None,
        "opened_at": None,
    }

    entries = _read_all(workspace_dir)
    entries.append(entry)
    _write_all(workspace_dir, entries)
    _note_path(workspace_dir, note_id).write_text(data.content)

    return NoteDetail(**entry, content=data.content)


def rename_note(workspace_id: str, note_id: str, title: str) -> Note:
    title = title.strip()
    if not title:
        raise HTTPException(status_code=422, detail="Note title cannot be empty")

    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["title"] = title
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    return _to_note(entry)


def update_content(workspace_id: str, note_id: str, content: str) -> Note:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    _note_path(workspace_dir, note_id).write_text(content)
    return _to_note(entry)


def set_flags(workspace_id: str, note_id: str, favorite: bool | None, pinned: bool | None) -> Note:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    if favorite is not None:
        entry["favorite"] = favorite
    if pinned is not None:
        entry["pinned"] = pinned
    _write_all(workspace_dir, entries)
    return _to_note(entry)


def set_folder(workspace_id: str, note_id: str, folder: str | None) -> Note:
    folder = folder.strip("/") if folder else None
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["folder"] = folder or None
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    return _to_note(entry)


def set_tags(workspace_id: str, note_id: str, tags: list[str]) -> Note:
    cleaned = sorted({t.strip() for t in tags if t.strip()})
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["tags"] = cleaned
    entry["updated_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)
    return _to_note(entry)


def duplicate_note(workspace_id: str, note_id: str) -> NoteDetail:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    source = _find(entries, note_id)
    content = _note_path(workspace_dir, note_id).read_text()

    now = datetime.now(UTC).isoformat()
    new_entry = {
        "id": uuid.uuid4().hex,
        "title": f"{source['title']} (copy)",
        "created_at": now,
        "updated_at": now,
        "favorite": False,
        "pinned": False,
        "folder": source["folder"],
        "tags": list(source["tags"]),
        "trashed": False,
        "trashed_at": None,
        "opened_at": None,
    }
    entries.append(new_entry)
    _write_all(workspace_dir, entries)
    _note_path(workspace_dir, new_entry["id"]).write_text(content)

    return NoteDetail(**new_entry, content=content)


def move_note(workspace_id: str, note_id: str, target_workspace_id: str) -> Note:
    if target_workspace_id == workspace_id:
        raise HTTPException(status_code=422, detail="Note is already in that workspace")

    workspace_dir = get_workspace_dir(workspace_id)
    target_dir = get_workspace_dir(target_workspace_id)

    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entries.remove(entry)

    target_entries = _read_all(target_dir)
    target_entries.append(entry)

    shutil.move(str(_note_path(workspace_dir, note_id)), str(_note_path(target_dir, note_id)))
    _write_all(workspace_dir, entries)
    _write_all(target_dir, target_entries)

    return _to_note(entry)


def trash_note(workspace_id: str, note_id: str) -> None:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["trashed"] = True
    entry["trashed_at"] = datetime.now(UTC).isoformat()
    _write_all(workspace_dir, entries)


def restore_note(workspace_id: str, note_id: str) -> Note:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entry["trashed"] = False
    entry["trashed_at"] = None
    _write_all(workspace_dir, entries)
    return _to_note(entry)


def purge_note(workspace_id: str, note_id: str) -> None:
    workspace_dir = get_workspace_dir(workspace_id)
    entries = _read_all(workspace_dir)
    entry = _find(entries, note_id)
    entries.remove(entry)
    _write_all(workspace_dir, entries)
    _note_path(workspace_dir, note_id).unlink(missing_ok=True)

File: apps/api/src/services/noteService.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { readJsonl, writeJsonl } from '../utils/fileDb.js';
import { getWorkspaceDir } from './workspaceService.js';
import { Note, NoteCreate, NoteDetail, NoteListItem } from '@ticks/types';

const METADATA_FILENAME = 'notes.jsonl';
const SNIPPET_LENGTH = 140;

function metadataPath(workspaceDir: string): string {
  return path.join(workspaceDir, METADATA_FILENAME);
}

function notePath(workspaceDir: string, noteId: string): string {
  return path.join(workspaceDir, 'notes', `${noteId}.md`);
}

// In TS we want to export the _read_all function so drawingService can import it
export function readAllNotes(workspaceDir: string): any[] {
  return readJsonl(metadataPath(workspaceDir));
}

function writeAllNotes(workspaceDir: string, entries: any[]): void {
  writeJsonl(metadataPath(workspaceDir), entries);
}

function sortNotes(entries: any[]): any[] {
  // Sort by updated_at desc, then by pinned desc
  return [...entries]
    .sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime())
    .sort((a, b) => (b.pinned ? 1 : 0) - (a.pinned ? 1 : 0));
}

function findNote(entries: any[], noteId: string): any {
  const entry = entries.find(e => e.id === noteId);
  if (!entry) {
    throw { status: 404, message: 'Note not found' };
  }
  return entry;
}

function makeSnippet(content: string, query: string): string {
  let cleanContent = content.replace(/\s+/g, ' ');
  if (!cleanContent) return '';
  if (!query) return cleanContent.slice(0, SNIPPET_LENGTH);

  const idx = cleanContent.toLowerCase().indexOf(query.toLowerCase());
  if (idx === -1) return cleanContent.slice(0, SNIPPET_LENGTH);

  const start = Math.max(0, idx - Math.floor(SNIPPET_LENGTH / 3));
  const end = Math.min(cleanContent.length, start + SNIPPET_LENGTH);
  const prefix = start > 0 ? '…' : '';
  const suffix = end < cleanContent.length ? '…' : '';
  return `${prefix}${cleanContent.slice(start, end)}${suffix}`;
}

function toListItem(workspaceDir: string, entry: any, query = ''): NoteListItem {
  const filePath = notePath(workspaceDir, entry.id);
  const content = fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8') : '';
  return {
    ...entry,
    snippet: makeSnippet(content, query)
  };
}

export function noteExists(workspaceId: string, noteId: string): boolean {
  const workspaceDir = getWorkspaceDir(workspaceId);
  return readAllNotes(workspaceDir).some(e => e.id === noteId);
}

export function listNotes(workspaceId: string, favoriteOnly = false, pinnedOnly = false): NoteListItem[] {
  const workspaceDir = getWorkspaceDir(workspaceId);
  let entries = sortNotes(readAllNotes(workspaceDir).filter(e => !e.trashed));
  if (favoriteOnly) {
    entries = entries.filter(e => e.favorite);
  }
  if (pinnedOnly) {
    entries = entries.filter(e => e.pinned);
  }
  return entries.map(e => toListItem(workspaceDir, e));
}

export function searchNotes(workspaceId: string, query: string, favoriteOnly = false, pinnedOnly = false): NoteListItem[] {
  const trimmedQuery = query.trim();
  if (!trimmedQuery) {
    return listNotes(workspaceId, favoriteOnly, pinnedOnly);
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const lowerQuery = trimmedQuery.toLowerCase();
  const matches: any[] = [];

  for (const entry of readAllNotes(workspaceDir)) {
    if (entry.trashed) continue;
    if (favoriteOnly && !entry.favorite) continue;
    if (pinnedOnly && !entry.pinned) continue;

    if (entry.title.toLowerCase().includes(lowerQuery)) {
      matches.push(entry);
      continue;
    }

    const filePath = notePath(workspaceDir, entry.id);
    const content = fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8') : '';
    if (content.toLowerCase().includes(lowerQuery)) {
      matches.push(entry);
    }
  }

  return sortNotes(matches).map(e => toListItem(workspaceDir, e, trimmedQuery));
}

export function listRecent(workspaceId: string, limit = 20): NoteListItem[] {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir).filter(e => e.opened_at && !e.trashed);
  entries.sort((a, b) => new Date(b.opened_at).getTime() - new Date(a.opened_at).getTime());
  return entries.slice(0, limit).map(e => toListItem(workspaceDir, e));
}

export function listTrash(workspaceId: string): NoteListItem[] {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir).filter(e => e.trashed);
  entries.sort((a, b) => new Date(b.trashed_at).getTime() - new Date(a.trashed_at).getTime());
  return entries.map(e => toListItem(workspaceDir, e));
}

export function listFolders(workspaceId: string): string[] {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const folders = new Set<string>();
  
  // Folders from notes
  for (const entry of readAllNotes(workspaceDir)) {
    if (entry.folder) folders.add(entry.folder);
  }
  
  // Explicit folders
  const foldersPath = path.join(workspaceDir, 'folders.jsonl');
  if (fs.existsSync(foldersPath)) {
    for (const f of readJsonl(foldersPath)) {
      if (f.name) folders.add(f.name);
    }
  }
  
  return Array.from(folders).sort();
}

export function createFolder(workspaceId: string, folderName: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const foldersPath = path.join(workspaceDir, 'folders.jsonl');
  const folders = fs.existsSync(foldersPath) ? readJsonl(foldersPath) : [];
  if (!folders.some(f => f.name === folderName)) {
    folders.push({ name: folderName, created_at: new Date().toISOString() });
    writeJsonl(foldersPath, folders);
  }
}

export function deleteFolder(workspaceId: string, folderName: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const foldersPath = path.join(workspaceDir, 'folders.jsonl');
  if (fs.existsSync(foldersPath)) {
    const folders = readJsonl(foldersPath).filter(f => f.name !== folderName);
    writeJsonl(foldersPath, folders);
  }
}

export function listTags(workspaceId: string): string[] {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const tags = new Set<string>();
  for (const entry of readAllNotes(workspaceDir)) {
    if (entry.tags) {
      for (const t of entry.tags) tags.add(t);
    }
  }
  return Array.from(tags).sort();
}

export function getNote(workspaceId: string, noteId: string): NoteDetail {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.opened_at = new Date().toISOString();
  writeAllNotes(workspaceDir, entries);
  const filePath = notePath(workspaceDir, noteId);
  const content = fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8') : '';
  return { ...entry, content };
}

export function createNote(workspaceId: string, data: NoteCreate): NoteDetail {
  const title = data.title.trim();
  if (!title) {
    throw { status: 422, message: 'Note title cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const noteId = crypto.randomUUID().replace(/-/g, '');
  const now = new Date().toISOString();

  const entry = {
    id: noteId,
    title,
    created_at: now,
    updated_at: now,
    favorite: false,
    pinned: false,
    folder: null as string | null,
    tags: [] as string[],
    trashed: false,
    trashed_at: null as string | null,
    opened_at: null as string | null
  };

  const entries = readAllNotes(workspaceDir);
  entries.push(entry);
  writeAllNotes(workspaceDir, entries);

  const filePath = notePath(workspaceDir, noteId);
  fs.writeFileSync(filePath, data.content || '', 'utf8');

  return { ...entry, content: data.content || '' };
}

export function renameNote(workspaceId: string, noteId: string, title: string): Note {
  const trimmedTitle = title.trim();
  if (!trimmedTitle) {
    throw { status: 422, message: 'Note title cannot be empty' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.title = trimmedTitle;
  entry.updated_at = new Date().toISOString();
  writeAllNotes(workspaceDir, entries);
  return entry;
}

export function updateContent(workspaceId: string, noteId: string, content: string): Note {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.updated_at = new Date().toISOString();
  writeAllNotes(workspaceDir, entries);
  fs.writeFileSync(notePath(workspaceDir, noteId), content, 'utf8');
  return entry;
}

export function setFlags(workspaceId: string, noteId: string, favorite: boolean | null | undefined, pinned: boolean | null | undefined): Note {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  if (favorite !== undefined && favorite !== null) entry.favorite = favorite;
  if (pinned !== undefined && pinned !== null) entry.pinned = pinned;
  writeAllNotes(workspaceDir, entries);
  return entry;
}

export function setComments(workspaceId: string, noteId: string, comments: any[]): Note {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.comments = comments;
  writeAllNotes(workspaceDir, entries);
  return entry;
}

export function setFolder(workspaceId: string, noteId: string, folder: string | null | undefined): Note {
  const cleanFolder = folder ? folder.replace(/^\/+|\/+$/g, '').trim() : null;
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.folder = cleanFolder || null;
  entry.updated_at = new Date().toISOString();
  writeAllNotes(workspaceDir, entries);
  return entry;
}

export function setTags(workspaceId: string, noteId: string, tags: string[]): Note {
  const cleaned = Array.from(new Set(tags.map(t => t.trim()).filter(t => t.length > 0))).sort();
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.tags = cleaned;
  entry.updated_at = new Date().toISOString();
  writeAllNotes(workspaceDir, entries);
  return entry;
}

export function duplicateNote(workspaceId: string, noteId: string): NoteDetail {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const source = findNote(entries, noteId);
  const filePath = notePath(workspaceDir, noteId);
  const content = fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8') : '';

  const now = new Date().toISOString();
  const newId = crypto.randomUUID().replace(/-/g, '');
  const newEntry = {
    id: newId,
    title: `${source.title} (copy)`,
    created_at: now,
    updated_at: now,
    favorite: false,
    pinned: false,
    folder: source.folder,
    tags: [...source.tags],
    trashed: false,
    trashed_at: null as string | null,
    opened_at: null as string | null
  };

  entries.push(newEntry);
  writeAllNotes(workspaceDir, entries);
  fs.writeFileSync(notePath(workspaceDir, newId), content, 'utf8');

  return { ...newEntry, content };
}

export function moveNote(workspaceId: string, noteId: string, targetWorkspaceId: string): Note {
  if (targetWorkspaceId === workspaceId) {
    throw { status: 422, message: 'Note is already in that workspace' };
  }

  const workspaceDir = getWorkspaceDir(workspaceId);
  const targetDir = getWorkspaceDir(targetWorkspaceId);

  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  const index = entries.indexOf(entry);
  if (index > -1) {
    entries.splice(index, 1);
  }

  const targetEntries = readAllNotes(targetDir);
  targetEntries.push(entry);

  const srcPath = notePath(workspaceDir, noteId);
  const destPath = notePath(targetDir, noteId);

  fs.renameSync(srcPath, destPath);
  writeAllNotes(workspaceDir, entries);
  writeAllNotes(targetDir, targetEntries);

  return entry;
}

export function trashNote(workspaceId: string, noteId: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.trashed = true;
  entry.trashed_at = new Date().toISOString();
  writeAllNotes(workspaceDir, entries);
}

export function restoreNote(workspaceId: string, noteId: string): Note {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = findNote(entries, noteId);
  entry.trashed = false;
  entry.trashed_at = null;
  writeAllNotes(workspaceDir, entries);
  return entry;
}

export function purgeNote(workspaceId: string, noteId: string): void {
  const workspaceDir = getWorkspaceDir(workspaceId);
  const entries = readAllNotes(workspaceDir);
  const entry = entries.find(e => e.id === noteId);
  if (entry) {
    const index = entries.indexOf(entry);
    if (index > -1) {
      entries.splice(index, 1);
    }
    writeAllNotes(workspaceDir, entries);
  }
  const filePath = notePath(workspaceDir, noteId);
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
  }
}
